relayer/message_coordinator.go

```diff
@@ -0,0 +1,287 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package relayer
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/database"
+	"github.com/ava-labs/awm-relayer/messages"
+	relayerTypes "github.com/ava-labs/awm-relayer/types"
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/ethclient"
+	"github.com/ethereum/go-ethereum/common"
+	"go.uber.org/zap"
+)
+
+var globalMessageCoordinator *MessageCoordinator
```

Comment:
I think we've generally avoided global vars, @cam-schultz is that correct? My personal preference has been to avoid them.

If we do decide to go with them though, should we have some checks that these are singleton instances?

Reply:
Michael, Cam, and I landed on having a global object for these shared resources. With the current design, `ApplicationRelayer`s are inherently global, and we need some way to access them. We can't access the right one without the `MessageHandlerFactories`.

We _could_ make this is singleton, but I don't really see a benefit of enforcing only one instantiation of it (though I don't think we _should_ instantiate more than 1)

Reply:
was just thinking to have a builtin check that makes sure `globalMessageCoordinator` isn't overrode with another instance

Reply:
It's a semantic difference, but we have a bunch of variables that live at `package main` scope and are passed around, as opposed to static global vars that live at application scope. We confusingly use the the term "global" for the former. 

They're functionally the same, but in order to remain consistent, we should return the `MessageCoordinator` instance to `main` and leave it to `main` to ensure that only one instance exists. We can then pass the `MessageCoordinator` reference to all of the downstream objects that interact with it, similar to how we treat the `RelayerDatabase` and `AppRequestNetwork` global singletons.